core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -271,31 +272,22 @@ class TransactionStateManager(brokerId: Int,
    * When this broker becomes a follower for a transaction log partition, clear out the cache for corresponding transactional ids
    * that belong to that partition.
    */
-  def removeTransactionsForPartition(partition: Int) {
+  def removeTransactionsForTxnTopicPartition(partitionId: Int) {
     validateTransactionTopicPartitionCountIsStable()
 
-    val topicPartition = new TopicPartition(Topic.TransactionStateTopicName, partition)
-
-    inLock(stateLock) {
-      ownedPartitions.remove(partition)
-      loadingPartitions.remove(partition)
-    }
+    val topicPartition = new TopicPartition(Topic.TransactionStateTopicName, partitionId)
 
     def removeTransactions() {
-      var numTxnsRemoved = 0
-
       inLock(stateLock) {
-        for (transactionalId <- transactionMetadataCache.keys) {
-          if (partitionFor(transactionalId) == partition) {
-            // we do not need to worry about whether the transactional id has any ongoing transaction or not since
-            // the new leader will handle it
-            transactionMetadataCache.remove(transactionalId)
-            numTxnsRemoved += 1
-          }
+        transactionMetadataCache.remove(partitionId) match {
+          case Some(txnMetadataCacheEntry) =>
+            info(s"Removed ${txnMetadataCacheEntry.metadataPerTransactionalId.size} cached transaction metadata for $topicPartition on follower transition")
+
+          case None =>
+            throw new IllegalStateException(s"The metadata cache entry for txn partition $partitionId does not exist while trying to remove it.")
```

Comment:
Hmm, that is a good point. I think it can happen that `onFollower` called twice but the second call's triggered thread gets executed before the first call. So yeah we could just log it and move on.